<a href="https://colab.research.google.com/github/atikhasan007/Scikit-learn/blob/main/Lasso(L1)_and_Ridge(L2)_Regularization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h2 style='color:blue' align='center'>Lasso(L1) and Ridge(L2) Regularization</h2>

In [ ]:
# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

In [ ]:
# Suppress Warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# read dataset
dataset = pd.read_csv('Melbourne_housing_FULL.csv')

In [ ]:
dataset

Suburb             Address  Rooms Type      Price Method  \
0      Abbotsford       68 Studley St      2    h        NaN     SS   
1      Abbotsford        85 Turner St      2    h  1480000.0      S   
2      Abbotsford     25 Bloomburg St      2    h  1035000.0      S   
3      Abbotsford  18/659 Victoria St      3    u        NaN     VB   
4      Abbotsford        5 Charles St      3    h  1465000.0     SP   
...           ...                 ...    ...  ...        ...    ...   
34852  Yarraville         13 Burns St      4    h  1480000.0     PI   
34853  Yarraville       29A Murray St      2    h   888000.0     SP   
34854  Yarraville      147A Severn St      2    t   705000.0      S   
34855  Yarraville    12/37 Stephen St      3    h  1140000.0     SP   
34856  Yarraville    3 Tarrengower St      2    h  1020000.0     PI   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis   3/09/2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds   4/02/2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
34852            Jas  24/02/2018       6.3    3013.0  ...       1.0  3.0   
34853        Sweeney  24/02/2018       6.3    3013.0  ...       2.0  1.0   
34854            Jas  24/02/2018       6.3    3013.0  ...       1.0  2.0   
34855  hockingstuart  24/02/2018       6.3    3013.0  ...       NaN  NaN   
34856             RW  24/02/2018       6.3    3013.0  ...       1.0  0.0   

       Landsize  BuildingArea  YearBuilt               CouncilArea Lattitude  \
0         126.0           NaN        NaN        Yarra City Council -37.80140   
1         202.0           NaN        NaN        Yarra City Council -37.79960   
2         156.0          79.0     1900.0        Yarra City Council -37.80790   
3           0.0           NaN        NaN        Yarra City Council -37.81140   
4         134.0         150.0     1900.0        Yarra City Council -37.80930   
...         ...           ...        ...                       ...       ...   
34852     593.0           NaN        NaN  Maribyrnong City Council -37.81053   
34853      98.0         104.0     2018.0  Maribyrnong City Council -37.81551   
34854     220.0         120.0     2000.0  Maribyrnong City Council -37.82286   
34855       NaN           NaN        NaN  Maribyrnong City Council       NaN   
34856     250.0         103.0     1930.0  Maribyrnong City Council -37.81810   

       Longtitude             Regionname Propertycount  
0       144.99580  Northern Metropolitan        4019.0  
1       144.99840  Northern Metropolitan        4019.0  
2       144.99340  Northern Metropolitan        4019.0  
3       145.01160  Northern Metropolitan        4019.0  
4       144.99440  Northern Metropolitan        4019.0  
...           ...                    ...           ...  
34852   144.88467   Western Metropolitan        6543.0  
34853   144.88826   Western Metropolitan        6543.0  
34854   144.87856   Western Metropolitan        6543.0  
34855         NaN   Western Metropolitan        6543.0  
34856   144.89351   Western Metropolitan        6543.0  

[34857 rows x 21 columns]

In [ ]:
dataset.nunique()

In [ ]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount',
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]

In [ ]:
dataset

Suburb  Rooms Type Method        SellerG             Regionname  \
0      Abbotsford      2    h     SS         Jellis  Northern Metropolitan   
1      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
2      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
3      Abbotsford      3    u     VB         Rounds  Northern Metropolitan   
4      Abbotsford      3    h     SP         Biggin  Northern Metropolitan   
...           ...    ...  ...    ...            ...                    ...   
34852  Yarraville      4    h     PI            Jas   Western Metropolitan   
34853  Yarraville      2    h     SP        Sweeney   Western Metropolitan   
34854  Yarraville      2    t      S            Jas   Western Metropolitan   
34855  Yarraville      3    h     SP  hockingstuart   Western Metropolitan   
34856  Yarraville      2    h     PI             RW   Western Metropolitan   

       Propertycount  Distance               CouncilArea  Bedroom2  Bathroom  \
0             4019.0       2.5        Yarra City Council       2.0       1.0   
1             4019.0       2.5        Yarra City Council       2.0       1.0   
2             4019.0       2.5        Yarra City Council       2.0       1.0   
3             4019.0       2.5        Yarra City Council       3.0       2.0   
4             4019.0       2.5        Yarra City Council       3.0       2.0   
...              ...       ...                       ...       ...       ...   
34852         6543.0       6.3  Maribyrnong City Council       4.0       1.0   
34853         6543.0       6.3  Maribyrnong City Council       2.0       2.0   
34854         6543.0       6.3  Maribyrnong City Council       2.0       1.0   
34855         6543.0       6.3  Maribyrnong City Council       NaN       NaN   
34856         6543.0       6.3  Maribyrnong City Council       2.0       1.0   

       Car  Landsize  BuildingArea      Price  
0      1.0     126.0           NaN        NaN  
1      1.0     202.0           NaN  1480000.0  
2      0.0     156.0          79.0  1035000.0  
3      1.0       0.0           NaN        NaN  
4      0.0     134.0         150.0  1465000.0  
...    ...       ...           ...        ...  
34852  3.0     593.0           NaN  1480000.0  
34853  1.0      98.0         104.0   888000.0  
34854  2.0     220.0         120.0   705000.0  
34855  NaN       NaN           NaN  1140000.0  
34856  0.0     250.0         103.0  1020000.0  

[34857 rows x 15 columns]

#### Checking for Nan values

In [ ]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

#### Handling Missing values

In [ ]:
cols_to_fill_zero = ['Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

In [ ]:
dataset

Suburb  Rooms Type Method        SellerG             Regionname  \
0      Abbotsford      2    h     SS         Jellis  Northern Metropolitan   
1      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
2      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
3      Abbotsford      3    u     VB         Rounds  Northern Metropolitan   
4      Abbotsford      3    h     SP         Biggin  Northern Metropolitan   
...           ...    ...  ...    ...            ...                    ...   
34852  Yarraville      4    h     PI            Jas   Western Metropolitan   
34853  Yarraville      2    h     SP        Sweeney   Western Metropolitan   
34854  Yarraville      2    t      S            Jas   Western Metropolitan   
34855  Yarraville      3    h     SP  hockingstuart   Western Metropolitan   
34856  Yarraville      2    h     PI             RW   Western Metropolitan   

       Propertycount  Distance               CouncilArea  Bedroom2  Bathroom  \
0             4019.0       2.5        Yarra City Council       2.0       1.0   
1             4019.0       2.5        Yarra City Council       2.0       1.0   
2             4019.0       2.5        Yarra City Council       2.0       1.0   
3             4019.0       2.5        Yarra City Council       3.0       2.0   
4             4019.0       2.5        Yarra City Council       3.0       2.0   
...              ...       ...                       ...       ...       ...   
34852         6543.0       6.3  Maribyrnong City Council       4.0       1.0   
34853         6543.0       6.3  Maribyrnong City Council       2.0       2.0   
34854         6543.0       6.3  Maribyrnong City Council       2.0       1.0   
34855         6543.0       6.3  Maribyrnong City Council       0.0       0.0   
34856         6543.0       6.3  Maribyrnong City Council       2.0       1.0   

       Car  Landsize  BuildingArea      Price  
0      1.0     126.0           NaN        NaN  
1      1.0     202.0           NaN  1480000.0  
2      0.0     156.0          79.0  1035000.0  
3      1.0       0.0           NaN        NaN  
4      0.0     134.0         150.0  1465000.0  
...    ...       ...           ...        ...  
34852  3.0     593.0           NaN  1480000.0  
34853  1.0      98.0         104.0   888000.0  
34854  2.0     220.0         120.0   705000.0  
34855  0.0       NaN           NaN  1140000.0  
34856  0.0     250.0         103.0  1020000.0  

[34857 rows x 15 columns]

In [ ]:
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

In [ ]:
dataset

Suburb  Rooms Type Method        SellerG             Regionname  \
0      Abbotsford      2    h     SS         Jellis  Northern Metropolitan   
1      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
2      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
3      Abbotsford      3    u     VB         Rounds  Northern Metropolitan   
4      Abbotsford      3    h     SP         Biggin  Northern Metropolitan   
...           ...    ...  ...    ...            ...                    ...   
34852  Yarraville      4    h     PI            Jas   Western Metropolitan   
34853  Yarraville      2    h     SP        Sweeney   Western Metropolitan   
34854  Yarraville      2    t      S            Jas   Western Metropolitan   
34855  Yarraville      3    h     SP  hockingstuart   Western Metropolitan   
34856  Yarraville      2    h     PI             RW   Western Metropolitan   

       Propertycount  Distance               CouncilArea  Bedroom2  Bathroom  \
0             4019.0       2.5        Yarra City Council       2.0       1.0   
1             4019.0       2.5        Yarra City Council       2.0       1.0   
2             4019.0       2.5        Yarra City Council       2.0       1.0   
3             4019.0       2.5        Yarra City Council       3.0       2.0   
4             4019.0       2.5        Yarra City Council       3.0       2.0   
...              ...       ...                       ...       ...       ...   
34852         6543.0       6.3  Maribyrnong City Council       4.0       1.0   
34853         6543.0       6.3  Maribyrnong City Council       2.0       2.0   
34854         6543.0       6.3  Maribyrnong City Council       2.0       1.0   
34855         6543.0       6.3  Maribyrnong City Council       0.0       0.0   
34856         6543.0       6.3  Maribyrnong City Council       2.0       1.0   

       Car    Landsize  BuildingArea      Price  
0      1.0  126.000000      160.2564        NaN  
1      1.0  202.000000      160.2564  1480000.0  
2      0.0  156.000000       79.0000  1035000.0  
3      1.0    0.000000      160.2564        NaN  
4      0.0  134.000000      150.0000  1465000.0  
...    ...         ...           ...        ...  
34852  3.0  593.000000      160.2564  1480000.0  
34853  1.0   98.000000      104.0000   888000.0  
34854  2.0  220.000000      120.0000   705000.0  
34855  0.0  593.598993      160.2564  1140000.0  
34856  0.0  250.000000      103.0000  1020000.0  

[34857 rows x 15 columns]

**Drop NA values of Price, since it's our predictive variable we won't impute it**

In [ ]:
dataset.isna().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Regionname          3
Propertycount       3
Distance            1
CouncilArea         3
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
Price            7610
dtype: int64

In [ ]:
dataset.dropna(inplace=True)

In [ ]:
dataset.isna().sum()

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Regionname       0
Propertycount    0
Distance         0
CouncilArea      0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
Price            0
dtype: int64

#### Let's one hot encode the categorical features

In [ ]:
dataset

Suburb  Rooms Type Method        SellerG             Regionname  \
1      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
2      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
4      Abbotsford      3    h     SP         Biggin  Northern Metropolitan   
5      Abbotsford      3    h     PI         Biggin  Northern Metropolitan   
6      Abbotsford      4    h     VB         Nelson  Northern Metropolitan   
...           ...    ...  ...    ...            ...                    ...   
34852  Yarraville      4    h     PI            Jas   Western Metropolitan   
34853  Yarraville      2    h     SP        Sweeney   Western Metropolitan   
34854  Yarraville      2    t      S            Jas   Western Metropolitan   
34855  Yarraville      3    h     SP  hockingstuart   Western Metropolitan   
34856  Yarraville      2    h     PI             RW   Western Metropolitan   

       Propertycount  Distance               CouncilArea  Bedroom2  Bathroom  \
1             4019.0       2.5        Yarra City Council       2.0       1.0   
2             4019.0       2.5        Yarra City Council       2.0       1.0   
4             4019.0       2.5        Yarra City Council       3.0       2.0   
5             4019.0       2.5        Yarra City Council       3.0       2.0   
6             4019.0       2.5        Yarra City Council       3.0       1.0   
...              ...       ...                       ...       ...       ...   
34852         6543.0       6.3  Maribyrnong City Council       4.0       1.0   
34853         6543.0       6.3  Maribyrnong City Council       2.0       2.0   
34854         6543.0       6.3  Maribyrnong City Council       2.0       1.0   
34855         6543.0       6.3  Maribyrnong City Council       0.0       0.0   
34856         6543.0       6.3  Maribyrnong City Council       2.0       1.0   

       Car    Landsize  BuildingArea      Price  
1      1.0  202.000000      160.2564  1480000.0  
2      0.0  156.000000       79.0000  1035000.0  
4      0.0  134.000000      150.0000  1465000.0  
5      1.0   94.000000      160.2564   850000.0  
6      2.0  120.000000      142.0000  1600000.0  
...    ...         ...           ...        ...  
34852  3.0  593.000000      160.2564  1480000.0  
34853  1.0   98.000000      104.0000   888000.0  
34854  2.0  220.000000      120.0000   705000.0  
34855  0.0  593.598993      160.2564  1140000.0  
34856  0.0  250.000000      103.0000  1020000.0  

[27244 rows x 15 columns]

In [ ]:
dataset = pd.get_dummies(dataset, drop_first=True)

In [ ]:
dataset

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car    Landsize  \
1          2         4019.0       2.5       2.0       1.0  1.0  202.000000   
2          2         4019.0       2.5       2.0       1.0  0.0  156.000000   
4          3         4019.0       2.5       3.0       2.0  0.0  134.000000   
5          3         4019.0       2.5       3.0       2.0  1.0   94.000000   
6          4         4019.0       2.5       3.0       1.0  2.0  120.000000   
...      ...            ...       ...       ...       ...  ...         ...   
34852      4         6543.0       6.3       4.0       1.0  3.0  593.000000   
34853      2         6543.0       6.3       2.0       2.0  1.0   98.000000   
34854      2         6543.0       6.3       2.0       1.0  2.0  220.000000   
34855      3         6543.0       6.3       0.0       0.0  0.0  593.598993   
34856      2         6543.0       6.3       2.0       1.0  0.0  250.000000   

       BuildingArea      Price  Suburb_Aberfeldie  ...  \
1          160.2564  1480000.0              False  ...   
2           79.0000  1035000.0              False  ...   
4          150.0000  1465000.0              False  ...   
5          160.2564   850000.0              False  ...   
6          142.0000  1600000.0              False  ...   
...             ...        ...                ...  ...   
34852      160.2564  1480000.0              False  ...   
34853      104.0000   888000.0              False  ...   
34854      120.0000   705000.0              False  ...   
34855      160.2564  1140000.0              False  ...   
34856      103.0000  1020000.0              False  ...   

       CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                    False                              False   
2                                    False                              False   
4                                    False                              False   
5                                    False                              False   
6                                    False                              False   
...                                    ...                                ...   
34852                                False                              False   
34853                                False                              False   
34854                                False                              False   
34855                                False                              False   
34856                                False                              False   

       CouncilArea_Nillumbik Shire Council  \
1                                    False   
2                                    False   
4                                    False   
5                                    False   
6                                    False   
...                                    ...   
34852                                False   
34853                                False   
34854                                False   
34855                                False   
34856                                False   

       CouncilArea_Port Phillip City Council  \
1                                      False   
2                                      False   
4                                      False   
5                                      False   
6                                      False   
...                                      ...   
34852                                  False   
34853                                  False   
34854                                  False   
34855                                  False   
34856                                  False   

       CouncilArea_Stonnington City Council  \
1                                     False   
2                                     False   
4                                     False   
5                                     False   
6                                  

#### Let's bifurcate our dataset into train and test dataset

In [ ]:
X = dataset.drop('Price', axis=1)
y = dataset['Price']

In [ ]:
X

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car    Landsize  \
1          2         4019.0       2.5       2.0       1.0  1.0  202.000000   
2          2         4019.0       2.5       2.0       1.0  0.0  156.000000   
4          3         4019.0       2.5       3.0       2.0  0.0  134.000000   
5          3         4019.0       2.5       3.0       2.0  1.0   94.000000   
6          4         4019.0       2.5       3.0       1.0  2.0  120.000000   
...      ...            ...       ...       ...       ...  ...         ...   
34852      4         6543.0       6.3       4.0       1.0  3.0  593.000000   
34853      2         6543.0       6.3       2.0       2.0  1.0   98.000000   
34854      2         6543.0       6.3       2.0       1.0  2.0  220.000000   
34855      3         6543.0       6.3       0.0       0.0  0.0  593.598993   
34856      2         6543.0       6.3       2.0       1.0  0.0  250.000000   

       BuildingArea  Suburb_Aberfeldie  Suburb_Airport West  ...  \
1          160.2564              False                False  ...   
2           79.0000              False                False  ...   
4          150.0000              False                False  ...   
5          160.2564              False                False  ...   
6          142.0000              False                False  ...   
...             ...                ...                  ...  ...   
34852      160.2564              False                False  ...   
34853      104.0000              False                False  ...   
34854      120.0000              False                False  ...   
34855      160.2564              False                False  ...   
34856      103.0000              False                False  ...   

       CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                    False                              False   
2                                    False                              False   
4                                    False                              False   
5                                    False                              False   
6                                    False                              False   
...                                    ...                                ...   
34852                                False                              False   
34853                                False                              False   
34854                                False                              False   
34855                                False                              False   
34856                                False                              False   

       CouncilArea_Nillumbik Shire Council  \
1                                    False   
2                                    False   
4                                    False   
5                                    False   
6                                    False   
...                                    ...   
34852                                False   
34853                                False   
34854                                False   
34855                                False   
34856                                False   

       CouncilArea_Port Phillip City Council  \
1                                      False   
2                                      False   
4                                      False   
5                                      False   
6                                      False   
...                                      ...   
34852                                  False   
34853                                  False   
34854                                  False   
34855                                  False   
34856                                  False   

       CouncilArea_Stonnington City Council  \
1                                     False   
2                                     False   
4        

In [ ]:
y

1        1480000.0
2        1035000.0
4        1465000.0
5         850000.0
6        1600000.0
           ...    
34852    1480000.0
34853     888000.0
34854     705000.0
34855    1140000.0
34856    1020000.0
Name: Price, Length: 27244, dtype: float64

In [ ]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size = 0.3, random_state=2)

In [ ]:
train_X

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car    Landsize  \
12799      2         3593.0       4.4       2.0       1.0  1.0    0.000000   
477        3         4836.0       6.3       0.0       0.0  0.0  593.598993   
29182      3        16166.0      14.7       3.0       2.0  2.0  580.000000   
21339      3         8801.0       8.4       3.0       1.0  2.0  593.598993   
22275      3        11918.0       5.2       3.0       1.0  2.0  212.000000   
...      ...            ...       ...       ...       ...  ...         ...   
24177      4         2894.0      10.2       4.0       1.0  0.0  828.000000   
15171      1         4442.0      10.1       1.0       1.0  1.0    0.000000   
8585       3        21650.0      11.2       3.0       1.0  1.0  563.000000   
3351       3        11204.0       7.8       3.0       1.0  1.0  404.000000   
30337      2         9264.0       7.5       0.0       0.0  0.0  593.598993   

       BuildingArea  Suburb_Aberfeldie  Suburb_Airport West  ...  \
12799       62.0000              False                False  ...   
477        160.2564              False                False  ...   
29182      160.2564              False                False  ...   
21339      160.2564              False                False  ...   
22275      160.2564              False                False  ...   
...             ...                ...                  ...  ...   
24177      160.2564              False                False  ...   
15171      160.2564              False                False  ...   
8585       118.0000              False                False  ...   
3351       160.2564              False                False  ...   
30337      160.2564              False                False  ...   

       CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
12799                                False                              False   
477                                  False                              False   
29182                                False                              False   
21339                                False                              False   
22275                                False                               True   
...                                    ...                                ...   
24177                                False                              False   
15171                                False                              False   
8585                                 False                              False   
3351                                 False                              False   
30337                                False                              False   

       CouncilArea_Nillumbik Shire Council  \
12799                                False   
477                                  False   
29182                                False   
21339                                False   
22275                                False   
...                                    ...   
24177                                False   
15171                                False   
8585                                 False   
3351                                 False   
30337                                False   

       CouncilArea_Port Phillip City Council  \
12799                                  False   
477                                    False   
29182                                  False   
21339                                  False   
22275                                  False   
...                                      ...   
24177                                  False   
15171                                  False   
8585                                   False   
3351                                   False   
30337                                  False   

       CouncilArea_Stonnington City Council  \
12799                                 False   
477                                    True   
29182    

In [ ]:
test_X

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car    Landsize  \
32841      4         4497.0      15.2       4.0       1.0  1.0  573.000000   
10813      2         4380.0       5.1       2.0       1.0  1.0    0.000000   
7211       2         6821.0       2.3       2.0       1.0  1.0  215.000000   
17799      1         8870.0       7.0       1.0       1.0  1.0    0.000000   
3397       3        11204.0       7.8       0.0       0.0  0.0  593.598993   
...      ...            ...       ...       ...       ...  ...         ...   
20856      4         5336.0      15.5       4.0       2.0  4.0  593.598993   
33317      3         9264.0       7.5       3.0       1.0  1.0  440.000000   
25323      4        21650.0      12.0       4.0       1.0  1.0  736.000000   
4641       2        10412.0       9.2       0.0       0.0  0.0  593.598993   
11400      2         6482.0       7.5       2.0       2.0  1.0   78.000000   

       BuildingArea  Suburb_Aberfeldie  Suburb_Airport West  ...  \
32841      119.0000              False                False  ...   
10813      160.2564              False                False  ...   
7211         0.0000              False                False  ...   
17799       39.0000              False                False  ...   
3397       160.2564              False                False  ...   
...             ...                ...                  ...  ...   
20856      231.0000              False                False  ...   
33317      160.2564              False                False  ...   
25323      160.2564              False                False  ...   
4641       160.2564              False                False  ...   
11400      160.2564              False                False  ...   

       CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
32841                                False                              False   
10813                                False                              False   
7211                                 False                              False   
17799                                False                              False   
3397                                 False                              False   
...                                    ...                                ...   
20856                                False                              False   
33317                                False                              False   
25323                                False                              False   
4641                                 False                              False   
11400                                False                              False   

       CouncilArea_Nillumbik Shire Council  \
32841                                False   
10813                                False   
7211                                 False   
17799                                False   
3397                                 False   
...                                    ...   
20856                                False   
33317                                False   
25323                                False   
4641                                 False   
11400                                False   

       CouncilArea_Port Phillip City Council  \
32841                                  False   
10813                                   True   
7211                                   False   
17799                                  False   
3397                                   False   
...                                      ...   
20856                                  False   
33317                                  False   
25323                                  False   
4641                                   False   
11400                                  False   

       CouncilArea_Stonnington City Council  \
32841                                 False   
10813                                 False   
7211     

#### Let's train our Linear Regression Model on training dataset and check the accuracy on test set

In [ ]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
reg.fit(train_X, train_y)

LinearRegression()

In [ ]:
reg.score(test_X, test_y)

0.1385368316156511

In [ ]:
reg.coef_

array([ 2.64001655e+05,  4.92118905e+00, -4.64732744e+04, -8.22349031e+04,
        1.17151153e+05,  4.29703140e+04,  2.35173742e+00,  4.70024530e+02,
        2.61342069e+05, -4.61008827e+04, -1.32312259e+05,  2.15453995e+05,
        1.20133839e+05,  2.72683980e+05,  1.74702166e+05, -8.00201053e+04,
       -1.50967673e+05, -4.94461444e+04,  1.17711126e+05, -1.28880253e+05,
       -3.49349656e+04, -8.54261329e+03,  6.19157896e+04, -2.57258028e+05,
       -1.10868116e+05, -2.42391674e+05,  1.79856957e+05, -1.22395100e+05,
        2.04625284e+05,  3.78258911e+04,  1.46681364e+05,  2.32198627e+03,
        7.35395731e+04, -5.29198411e+04,  1.90482393e+05, -3.20451670e+05,
        7.49515014e+04, -2.83019804e+04,  2.69103300e+04,  2.78824270e+05,
        1.51386730e+05, -1.06324212e+05, -8.45087700e+04,  2.96568248e+05,
        1.50644588e+05,  6.84522092e-08,  1.75999576e+05, -1.20057166e+05,
       -9.64544651e+03,  3.68377378e+05, -1.35478625e+05, -6.38092126e+04,
        3.53493139e+04,  

In [ ]:
reg.score(train_X, train_y)

0.6827792395792723

**Here training score is 68% but test score is 13.85% which is very low**

<h4 style='color:purple'>Normal Regression is clearly overfitting the data, let's try other models</h4>

#### Using Lasso (L1 Regularized) Regression Model

In [ ]:
from sklearn import linear_model
#lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg = linear_model.Lasso(alpha=50, tol=0.1)
lasso_reg.fit(train_X, train_y)

Lasso(alpha=50, tol=0.1)

In [ ]:
lasso_reg.score(test_X, test_y)

0.6636280170612747

In [ ]:
lasso_reg.score(train_X, train_y)

0.6767149418617553

#### Using Ridge (L2 Regularized) Regression Model

In [ ]:
from sklearn.linear_model import Ridge
#ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg= Ridge(alpha=50, tol = 0.1)
ridge_reg.fit(train_X, train_y)

Ridge(alpha=50, tol=0.1)

In [ ]:
ridge_reg.score(test_X, test_y)

0.6670848945194956

In [ ]:
ridge_reg.score(train_X, train_y)

0.6622376739684328

In [ ]:
from sklearn import linear_model
#ElasticNet_reg = linear_model.ElasticNet(alpha=0.005, l1_ratio=0.9)
ElasticNet_reg = linear_model.ElasticNet(alpha=0.005, l1_ratio=0.9)
ElasticNet_reg.fit(train_X, train_y)

ElasticNet(alpha=0.005, l1_ratio=0.9)

In [ ]:
ElasticNet_reg.score(test_X, test_y)

0.6743976868352917

In [ ]:
ElasticNet_reg.score(train_X, train_y)

0.6747498827555971

**We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too**